File: src/net/merayen/elastic/ui/objects/top/views/arrangementview/Arrangement.java
package net.merayen.elastic.ui.objects.top.views.arrangementview;

import java.util.ArrayList;
import java.util.List;

import net.merayen.elastic.ui.UIObject;
import net.merayen.elastic.ui.objects.components.Button;
import net.merayen.elastic.ui.objects.components.Scroll;
import net.merayen.elastic.ui.objects.components.autolayout.AutoLayout;
import net.merayen.elastic.ui.objects.components.autolayout.LayoutMethods;

class Arrangement extends UIObject {
	float width, height;
	private TrackList track_list = new TrackList();
	private final Scroll arrangement_list_scroll = new Scroll(track_list);
	private AutoLayout button_bar = new AutoLayout(new LayoutMethods.HorizontalBox(5, 100000));
	private final List<UIObject> to_remove = new ArrayList<>();

	@Override
	protected void onInit() {
		Arrangement self = this;
		add(arrangement_list_scroll);
		add(button_bar);

		arrangement_list_scroll.translation.y = 20;

		button_bar.add(new Button() {{
			label = "New track";
			setHandler(() -> {
				Track track = new Track();
				track_list.add(track);
				track.setHandler(new Track.Handler() {
					@Override
					public void onRemove() {
						to_remove.add(track);
					}
				});
			});
		}});
	}

	@Override
	protected void onUpdate() {
		track_list.width = width;
		arrangement_list_scroll.width = width;
		arrangement_list_scroll.height = height - 20;

		for(UIObject obj : to_remove)
			track_list.remove(obj);
		to_remove.clear();
	}
}

File: src/net/merayen/elastic/ui/objects/top/views/arrangementview/Arrangement.kt
package net.merayen.elastic.ui.objects.top.views.arrangementview

import net.merayen.elastic.system.intercom.CreateNodeMessage
import net.merayen.elastic.ui.UIObject
import net.merayen.elastic.ui.objects.components.Button
import net.merayen.elastic.ui.objects.components.Scroll
import net.merayen.elastic.ui.objects.components.autolayout.AutoLayout
import net.merayen.elastic.ui.objects.components.autolayout.LayoutMethods
import net.merayen.elastic.util.Postmaster

internal class Arrangement : UIObject() {
	var layoutWidth: Float = 0f
	var layoutHeight: Float = 0f

	private val arrangementData = ArrangementData()

	private val trackList = TrackList()
	private val arrangementEventView = ArrangementEventView()
	private val arrangementListScroll = Scroll(arrangementEventView)
	private val buttonBar = AutoLayout(LayoutMethods.HorizontalBox(5f, 100000f))

	override fun onInit() {
		add(arrangementListScroll)
		add(buttonBar)
		add(trackList)

		trackList.translation.y = 20f

		arrangementListScroll.translation.x = 100f
		arrangementListScroll.translation.y = 20f

		buttonBar.add(object : Button() {
			init {
				label = "New track"
				handler = object : Button.IHandler {
					override fun onClick() {
						val track = Track()
						trackList.add(track)
						track.handler = object : Track.Handler {
							override fun onRemove() = trackList.remove(track)
						}
					}
				}
			}
		})
	}

	override fun onUpdate() {
		trackList.layoutWidth = layoutWidth
		arrangementListScroll.layoutWidth = layoutWidth
		arrangementListScroll.layoutHeight = layoutHeight - 20

		arrangementEventView.layoutWidth = layoutWidth - 100
		arrangementEventView.layoutHeight = layoutHeight - 20
	}

	fun handleMessage(message: Postmaster.Message) {
		when (message) {
			is CreateNodeMessage -> {
				if (message.name == "midi") {
					arrangementEventView.handleMessage(message)
				}
			}
		}
	}
}